### ISSUE TITLE
TypedDict subclasses should satisfy protocols requiring `__required_keys__` / `__optional_keys__` ClassVar members

### ISSUE THREAD

**Issue by 5hal1n** [astral-sh/ty#2826]
## Summary

ty reports `invalid-argument-type` when passing a `TypedDict` subclass to a function/constructor that expects a type satisfying a protocol with `ClassVar[frozenset[str]]` members like `__required_keys__` and `__optional_keys__`. Python's `TypedDict` metaclass provides these attributes at runtime, so the TypedDict subclass does satisfy the protocol — but ty cannot verify this statically.

## Reproduction

```python
from typing import ClassVar, Protocol, TypedDict

class TypedDictLikeV1(Protocol):
    __required_keys__: ClassVar[frozenset[str]]
    __optional_keys__: ClassVar[frozenset[str]]

class MyState(TypedDict):
    name: str
    value: int

def accepts_typeddict_like(schema: type[TypedDictLikeV1]) -> None: ...

accepts_typeddict_like(MyState)  # ty reports invalid-argument-type
```

**Expected**: No error — `MyState` has `__required_keys__` and `__optional_keys__` provided by the TypedDict metaclass.

**Actual**: ty reports:
```
error[invalid-argument-type]: Argument to function `accepts_typeddict_like` is not assignable to parameter `schema` of type `type[TypedDictLikeV1]`
```

## Real-world impact

This pattern is used by [LangGraph](https://github.com/langchain-ai/langgraph)'s `StateGraph` class, which accepts `TypedDict` subclasses as the state schema via a [`StateLike`](https://github.com/langchain-ai/langgraph/blob/main/libs/langgraph/langgraph/_internal/_typing.py) / `TypedDictLikeV1` protocol:

```python
from langgraph.graph import StateGraph
from typing import TypedDict

class MyState(TypedDict):
    messages: list[str]

builder = StateGraph(MyState)  # invalid-argument-type
```

This forces users of LangGraph (a widely-used AI framework) to add `# ty: ignore[invalid-argument-type]` on every `StateGraph()` call.

See also: https://github.com/langchain-ai/langgraph/issues/6823

## Related

- #1380 — Support `ClassVar` protocol members (this is the underlying cause)

## Environment

- ty version: 0.0.17
- Python: 3.13
- LangGraph: 0.4.1

**Comment by carljm** [comment]
It seems clearly unsound to accept this, the way pyright/mypy do (pyrefly does not, they emit the same diagnostic we do). Pyright and mypy both accept this:

```py
from typing import ClassVar, Protocol, TypedDict

class TypedDictLikeV1(Protocol):
    __required_keys__: ClassVar[frozenset[str]]
    __optional_keys__: ClassVar[frozenset[str]]

class MyState(TypedDict):
    name: str
    value: int

def accepts_typeddict_like(schema: type[TypedDictLikeV1]) -> None:
    instance = schema()
    instance.__required_keys__

accepts_typeddict_like(MyState)
```

There are several unsound things accepted there. First of all, `schema()` is accepted even though we have no idea what the constructor signature is of an object that is valid for `type[TypedDictLikeV1]`. And secondly, `instance` is then inferred as of type `TypedDictLikeV1`, and pyright/mypy then believe that it has `__required_keys__`, which is not true at runtime.

It seems like the sound approach for LangGraph would be this version:

```py
from typing import Protocol, TypedDict

class TypedDictLikeV1(Protocol):
    __required_keys__: frozenset[str]
    __optional_keys__: frozenset[str]

class MyState(TypedDict):
    name: str
    value: int

def accepts_typeddict_like(schema: TypedDictLikeV1) -> None:
    ...

accepts_typeddict_like(MyState)
```

This works today in ty and pyright and pyrefly, but unfortunately mypy does not think `MyState` is assignable to `TypedDictLikeV1` -- not sure why.

**Comment by cbornet** [comment]
One of the ways LangGraph uses TypedDictLikeV1 is like this (simplified):

```python
from typing import Protocol, TypedDict, Generic, TypeVar, ClassVar

class TypedDictLikeV1(Protocol):
    __required_keys__: ClassVar[frozenset[str]]
    __optional_keys__: ClassVar[frozenset[str]]

class TypedDictLikeV2(Protocol):
    __required_keys__: frozenset[str]
    __optional_keys__: frozenset[str]

TypedDictLikeT = TypeVar("TypedDictLikeT", bound=TypedDictLikeV1 | TypedDictLikeV2)

class MyGeneric(Generic[TypedDictLikeT]):
    pass

class Foo(TypedDict):
    pass

def foo(foo: MyGeneric[Foo]) -> None:
    pass
```

This passes with mypy (thanks to ClasVar) but not with ty:
```
error[invalid-type-arguments]: Type `Foo` is not assignable to upper bound `TypedDictLikeV1 | TypedDictLikeV2` of type variable `TypedDictLikeT@MyGeneric`
   --> tests/unit_tests/chat_models/test_chat_models.py:371:24
    |
371 | def foo(foo: MyGeneric[Foo]) -> None:
    |                        ^^^
    |
   ::: tests/unit_tests/chat_models/test_chat_models.py:363:1
    |
363 | TypedDictLikeT = TypeVar("TypedDictLikeT", bound=TypedDictLikeV1 | TypedDictLikeV2)
    | -------------- Type variable defined here
    |
info: type `Foo` is not assignable to any element of the union `TypedDictLikeV1 | TypedDictLikeV2`
info: ├── type `Foo` is not assignable to protocol `TypedDictLikeV1`
info: │   └── protocol member `__optional_keys__` is not defined on type `Foo`
info: └── type `Foo` is not assignable to protocol `TypedDictLikeV2`
info:     └── protocol member `__optional_keys__` is not defined on type `Foo`
```

How can we write it so that it passes with ty ?

**Comment by carljm** [comment]
This runs into exactly the same unsoundness discussed above. The runtime objects that inhabit the type `Foo` are regular dictionaries and do not have `__required_keys__` or `__optional_keys__` attributes. The `ClassVar` annotations do not help -- class variables are accessible via instances in Python, and that's how they are modeled in the Python type system, so those annotations still promise that these attributes will be available on instances of that type.

So `MyGeneric` is annotated as requiring that every instance of its contained type will have those two attributes, and that is not true for instances of `Foo`, therefore `MyGeneric[Foo]` is not a valid specialization of `MyGeneric`.

My best recommendation is to stop trying to distinguish typed-dicts from regular dicts at all in instance contexts (e.g. by using `Mapping[str, object]` as the upper bound instead); this distinction doesn't make sense, since TypedDict instances at runtime are just regular dicts, and thus there's no sound way to express it. (As shown in my last comment, you can use such a protocol to define an API that a TypedDict _class object_ can be passed to, because the class object really does have those attributes).
